content/4_Gold/Ternary_Search.mdx

```diff
@@ -0,0 +1,142 @@
+---
+id: ternary-search
+title: Ternary Search
+author: Chongtian Ma
+prerequisites:
+  - /silver/binary-search
+description: Searching for a peak in a unimodal function.
+frequency: 1
+---
+
+<Resources>
+	<Resource
+		source="CP-Algorithms"
+		title="Ternary Search"
+		url="https://cp-algorithms.com/num_methods/ternary_search.html"
+		starred
+	/>
+</Resources>
+
+Ternary search is the equivalent of binary searching for the zero on the
+derivative function. The most basic function can be represented by the standard
+form of a parabola: $ax^2+bx+c$. Depending on if $a$ is positive or negative,
+you are searching for the local minimum or maximum of the function,
+respectively. Note that the function must be unimodal (i.e. change directions in exactly one place)!
+
+Unlike binary search where a checking function $f(x)$ only needs to return true
+or false, we need $f(x)$ to output an integer. Between the bounds of $[l, r]$,
+we need to compare the integer output of $f(m_1)$ with $f(m_2)$ where
+$l < m_1 < m_2 < r$ to determine how $l$ and $r$ should be moved.
+
+For example, if the function has a local maximum between $[l, r]$ and
+$f(m_1) < f(m_2)$, then we know the local maximum cannot be between $[l, m_1]$,
+so we continue searching in $[m_1, r]$. For a mini-proof, consider the following
+cases:
+
+- Let's denote the local maximum as $g$. If $m_1$ and $m_2$ are less than $g$,
+  this means that $f(m_1)$ and $f(m_2)$ are both on the increasing side of the
+  function. Therefore, we know $m_2 < g < r$, so we must keep searching between
+  $[m_2, r]$.
+
+- If $m_1 < g$ and $m_2 > g$, then $m_1 < g < m_2$. Therefore, we must keep
+  searching betwen $[m_1, m_2]$.
+
+- Note that $g < m_1 < m_2$ can never happen, becuase the function had to be
+  strictly decreasing after $g$.
+
+Since $[m_1, m_2] \cup [m_2, r] = [m_1, r]$, we cover all cases if we set $l$ to
+$m_1$. A similar proof can be given for $f(m_1) \geq f(m_2)$ or if we are
+searching for a local minimum.
+
+It is usually optimal to set $m_1$ and $m_2$ equidistant to each other and to
+the endpoints, so we can maximize the amount of distance we eliminate per
+search. We can set $m_1 = l + \frac{r-l}{3}$ and $m_2 = r - \frac{r-l}{3}$ so
+the intervals $[l, m_1], [m_1, m_2],$ and $[m_2,r]$ have roughly equal length.
+
+# Sample Problem - Haybale Distribution
+
+<FocusProblem problem="sample" />
+
+## Solution - Haybale Distribution
+
+Consider the function $f(y)$, which returns the number of haybales wasted if $y$
+is the chosen integer point. The problem asks for the local minimum for $f(y)$
+in the range of integers $[0, 10^6]$. It can be shown the function is unimodal
+because:
+
+- If we assume $y'$ is the optimal value of $y$ (i.e. the local minimum), it
+  holds that $f(x) \leq f(x+1)$ for all $x \in [0, y'-1]$. Similarily,
```

Comment:
allowing equality isn't sufficient to ternary search. For example, consider a function that is $0$ at a point and $1$ everywhere else. you can't efficiently find the $0$ by ternary searching.

Reply:
is it better now?

Reply:
![image](https://github.com/user-attachments/assets/4e6a0cdb-d96c-4964-8f05-e1c7db90244e)

Reply:
you are just restating the definition of unimodality here. but why is it unimodal?

Reply:
do you want a rigorous proof in the context of the problem?

Reply:
> Is there really no other simpler way?

I don't think so? what's not simple about it?

maybe earlier in the module you can define unimodality and convexity in a way such that convexity obviously implies unimodality. E.g., for functions on which you want to find a local minimum,
- unimodality -> sign of slope is non-decreasing
- convex -> slope is non-decreasing